velox/vector/ComplexVector.cpp

```diff
@@ -809,31 +809,58 @@ VectorPtr RowVector::pushDictionaryToRowVectorLeaves(const VectorPtr& input) {
       wrappers, input->size(), input, input->pool());
 }
 
-void ArrayVectorBase::checkRanges() const {
-  std::unordered_map<vector_size_t, vector_size_t> seenElements;
-  seenElements.reserve(size());
+template <bool kHasNulls>
+vector_size_t ArrayVectorBase::nextNonEmpty(vector_size_t i) const {
+  while (i < size() &&
+         ((kHasNulls && bits::isBitNull(rawNulls(), i)) || rawSizes_[i] <= 0)) {
+    ++i;
+  }
+  return i;
+}
+
+template <bool kHasNulls>
+bool ArrayVectorBase::maybeHaveOverlappingRanges() const {
+  vector_size_t i = 0;
+  i = nextNonEmpty<kHasNulls>(i);
+  if (i >= size()) {
+    return false;
+  }
+  for (;;) {
+    auto j = nextNonEmpty<kHasNulls>(i + 1);
+    if (j >= size()) {
+      return false;
+    }
+    if (rawOffsets_[i] + rawSizes_[i] > rawOffsets_[j]) {
```

Comment:
Here we also ensure that rawOffsets_ are in ascending order and not just overlapping? So we can explicit comment that also? Thanks!